### PR TITLE
Test sriov connectivity in functional suite

### PR DESCRIPTION
**Pull request by booxter**
**What this PR does / why we need it**: test sriov connectivity

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by booxter** [comment]
/hold

We also need to adjust the kubevirtci scripts to enforce VF link state up before it works on unplugged PFs (of compatible drivers that support in-PF switching).

File: tests/vmi_multus_test.go [hunk 1]
```diff
@@ -646,10 +653,10 @@ var _ = Describe("SRIOV", func() {
 
 			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
-			tests.WaitUntilVMIReady(vmi, tests.LoggedInFedoraExpecter)
 
 			// Need to wait for cloud init to finish and start the agent inside the vmi.
 			tests.WaitAgentConnected(virtClient, vmi)
+			tests.WaitUntilVMIReady(vmi, tests.LoggedInFedoraExpecter)
```
Comment (fedepaol) [inline tests/vmi_multus_test.go]: Why did you need to move this?

File: tests/vmi_multus_test.go [hunk 2]
```diff
@@ -44,11 +44,12 @@ import (
...
+	ptpConfWithTuningCRD   = `{"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"name":"%s","namespace":"%s"},"spec":{"config":"{ \"cniVersion\": \"0.3.1\", \"name\": \"mynet\", \"plugins\": [{\"type\": \"ptp\", \"ipam\": { \"type\": \"host-local\", \"subnet\": \"10.1.1.0/24\" }},{\"type\": \"tuning\"}]}"}}`
+	sriovConfCRD           = `{"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"name":"%s","namespace":"%s","annotations":{"k8s.v1.cni.cncf.io/resourceName":"%s"}},"spec":{"config":"{ \"name\": \"sriov\", \"type\": \"sriov\", \"ipam\": { \"type\": \"host-local\", \"subnet\": \"10.1.1.0/24\" } }"}}`
+	sriovLinkEnableConfCRD = `{"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"name":"%s","namespace":"%s","annotations":{"k8s.v1.cni.cncf.io/resourceName":"%s"}},"spec":{"config":"{ \"name\": \"sriov\", \"type\": \"sriov\", \"link_state\": \"enable\", \"ipam\": { \"type\": \"host-local\", \"subnet\": \"10.1.1.0/24\" } }"}}`
```
Comment (fedepaol) [inline tests/vmi_multus_test.go]: Would it make sense to have only one enabled CRD instead of the "enabled" one and the "regular" one?

Comment (booxter) [reply tests/vmi_multus_test.go]: The `enabled` version may not pass on older drivers that don't support the feature. Having both makes it slightly easier to skip the offending one in older environment. But I can change if you believe it's not enough of justification.

File: tests/vmi_multus_test.go (hunk 1 above)
Comment (booxter) [reply tests/vmi_multus_test.go]: @fedepaol I had this change split into a separate commit that explains the reason in a commit message. Strictly speaking, this commit can be merged separately, I just needed it to consistently pass the suite locally: https://github.com/kubevirt/kubevirt/pull/2692/commits/f96f646edd8a75ad6ceb2b88d952fc7c9857a5fd Would you like me to split it out?

Comment (fedepaol) [reply tests/vmi_multus_test.go]: @booxter no it's fine, sorry I did not read the commit message. All clear now.

Comment (mmirecki) [reply tests/vmi_multus_test.go]: Is it worth checking "WaitUntilVMIReady"?
If the agent is connected, then the vm must be up.

Comment (booxter) [reply tests/vmi_multus_test.go]: @mmirecki AFAIU we use the expecter to log in for next commands to succeed. Otherwise the next expecter checks will try to type in commands into `getty` login session.

**Comment by booxter** [comment]
The CNI plugin fix finally merged. Now we need to rebase this patch and bump the CNI version to include the link_state support to make this, hopefully, pass in upstream CI.

**Comment by booxter** [comment]
rebased; no explicit image bump needed because we already use `:latest` in `kubevirtci` repo.

**Comment by booxter** [comment]
@fedepaol @mmirecki yay, the new connectivity test case just passed the sriov lane.

**Comment by fedepaol** [comment]
> @fedepaol @mmirecki yay, the new connectivity test case just passed the sriov lane.

That's awesome! We should check it works on both nodes (one has intel, the other one is mellanox)

**Comment by booxter** [comment]
@fedepaol what's the procedure to land the job on a particular node (if any)? Or do we just /retest until it hits the right one?

**Comment by fedepaol** [comment]
@booxter I am not sure there is a way to see where the job landed.
I am able to schedule it manually on the node by changing the prow spec manually.
Will try tomorrow, just to have more control and not find out that the jobs are failing randomly because sometimes they get scheduled on the node that is not working.

**Comment by booxter** [comment]
@fedepaol isn't prow spec stored in a different repository? If so, how would you approach triggering a job on a particular node with an unmerged PR included?

**Comment by fedepaol** [comment]
> @fedepaol isn't prow spec stored in a different repository? If so, how would you approach triggering a job on a particular node with an unmerged PR included?

It's the same way I tested the kind lane before officially submitting it.
[Here](https://github.com/kubevirt/project-infra/tree/master/github/ci#testing-new-prowjobs) the instructions. You must have permissions over the prow cluster (which I have), and then it boils down to clone the actual prow job and changing the node selector to run on a specific node (instead of having the sr-iov selector).

**Comment by fedepaol** [comment]
This run https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/2692/pull-kubevirt-e2e-kind-k8s-sriov-1.14.2/1184384857543806976 is on pnx srv20

**Comment by fedepaol** [comment]
It failed, on pnx20 we have mellanox. 
This https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/2692/pull-kubevirt-e2e-kind-k8s-sriov-1.14.2/1184399579743784960 is running the same on pnx17

**Comment by booxter** [comment]
ftr I couldn't reproduce the same failure on another machine with a Mellanox PF. I probably need to compare NICs in gate and on that machine.

**Comment by booxter** [comment]
/retest

Let's see if we can reproduce it more than once.

**Comment by booxter** [comment]
Looks like the failure is consistent when and once the job lands on Mellanox machine. Note that I cannot reproduce the same on another Mellanox machine that I have in possession. The other machine has the exact same NIC (up to part number etc.) To confirm that firmware versions are identical. Also to extract from CI node: dmesg in case any sriov driver warnings logged there.

**Comment by booxter** [comment]
Looks like latest patches in the series broke all test cases, not just the new connectivity scenario. On all cards. Investigating.

**Comment by booxter** [comment]
I could reproduce the failure with AgentConnected locally once I set /proc/sys/net/bridge/bridge-nf-call-*tables to 0. Trying out if this is the issue in CI with a patch. (we'll have to merge it in cluster-up first if this is the issue.)

**Comment by booxter** [comment]
/test pull-kubevirt-e2e-kind-k8s-sriov-1.14.2

some deployment time issue happened

**Comment by booxter** [comment]
Looks like the sriov lane is fixed and now I got this error from the new test case:

```
/root/go/src/kubevirt.io/kubevirt/tests/vmi_multus_test.go:755
Failed to ping VMI testvmiqprnvxgpms2grt9ftskv7sfc52l6988glcmvhgzrk4gn99wg within the given timeout
Unexpected error:
    <expect.TimeoutError>: 30000000000
    expect: timer expired after 30 seconds
occurred
/root/go/src/kubevirt.io/kubevirt/tests/vmi_multus_test.go:850
```

from https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/2692/pull-kubevirt-e2e-kind-k8s-sriov-1.14.2/1222216818026876928

It's good we don't see AgentConnected issue anymore. This one is new. The failure suggests there is some issue with inter-port connectivity inside the PF switch. (Assuming PF even supports such switching, need to check somehow, perhaps dump more information about PFs used on deployment.)

**Comment by booxter** [comment]
The second run results in exactly the same failure. Need to check if that's an interface being down on one of VMIs or maybe in-PF switching not working.

**Comment by booxter** [comment]
@phoracek I believe this PR is finally nailed. The last issue was with how sriov CNI plugin leaves configured MAC addresses set for VFs and how that affects connectivity testing after duplicate MACs are introduced on VF links.

**Comment by booxter** [comment]
@phoracek it's green!!! https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/2692/pull-kubevirt-e2e-kind-k8s-sriov-1.14.2/1222560711008325632

**Comment by phoracek** [comment]
No way! Victory!

File: tests/vmi_multus_test.go [hunk 3]
```diff
@@ -775,6 +826,19 @@ func checkInterface(vmi *v1.VirtualMachineInstance, interfaceName, prompt string
 	Expect(err).ToNot(HaveOccurred(), "Interface %q was not found in the VMI %s within the given timeout", interfaceName, vmi.Name)
 }
 
+func configureSRIOV(vmi *v1.VirtualMachineInstance, interfaceName, addr, prompt string) {
```
Comment (phoracek) [inline tests/vmi_multus_test.go]: Can't you use `configInterface`?

**Comment by phoracek** [comment]
@booxter what I meant is that there already is a helper setting IP address few lines above the one you just introduced.

**Comment by booxter** [comment]
> @booxter what I meant is that there already is a helper setting IP address few lines above the one you just introduced.

Oh. I missed it exists. :)

**Comment by booxter** [comment]
And... the test case just failed again with the same error (which is quite generic so probably a different root cause). Hold on, need to investigate it again. This is so tiresome.

/hold

**Comment by booxter** [comment]
/hold cancel

There was a bug in how I used configInterface (not passing /24 mask) that broke the test. Now fixed.

**Comment by booxter** [comment]
@rmohr @dhiller @danielBelenky this is ready to merge, Petr already approved. Please take a look. Thanks.

File: tests/vmi_multus_test.go [hunk 4]
```diff
@@ -632,14 +641,15 @@ var _ = Describe("SRIOV", func() {
...
 
 		startVmi := func(vmi *v1.VirtualMachineInstance) {
-
 			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
-			tests.WaitUntilVMIReady(vmi, tests.LoggedInFedoraExpecter)
+			return
+		}
 
+		waitVmi := func(vmi *v1.VirtualMachineInstance) {
```
Comment (dhiller) [inline tests/vmi_multus_test.go]: nit: I'd prefer a more expressive name like `waitUntilVMIReady`
